### PR TITLE
AWS EKS CNI Monitoring Support

### LINKED ISSUE #271: AWS EKS CNI Monitoring support
**What is missing?**
When deploying `kube-prometheus` to AWS EKS, we cannot monitor the pod network related metrics ([awscni](https://github.com/aws/amazon-vpc-cni-k8s)), even when AWS CNI plugin has exposed [prometheus metrics](https://github.com/aws/amazon-vpc-cni-k8s/blob/c34df1016819af275d72858755700286fc94c03b/cni-metrics-helper/metrics/cni_metrics.go).

**Why do we need it?**
In order to monitor `available IP addresses` in any EKS Cluster & other awscni related metrics.

**Environment**
AWS EKS, any version.

* kube-prometheus version:

    `24aebaf98578d03b7ef85b2eb404260b3ce15f1a ` (latest master).

**Anything else we need to know?**:
N/A

### PR DESCRIPTION
**Pull request by karancode**
Resolves #271 
- Creates a `serviceMonitor` & the `service` for getting the metrics.

### CODE REVIEW

File: manifests/prometheus-rules.yaml [hunk 1]
```diff
@@ -79,6 +79,22 @@ spec:
...
+        * on (namespace, pod) group_left(node) max by(namespace, pod, node) (kube_pod_info)
+      record: node_namespace_pod_container:container_memory_rss
+    - expr: |
+        container_memory_cache{job="kubelet", image!=""}
+        * on (namespace, pod) group_left(node) max by(namespace, pod, node) (kube_pod_info)
+      record: node_namespace_pod_container:container_memory_cache
+    - expr: |
+        container_memory_swap{job="kubelet", image!=""}
+        * on (namespace, pod) group_left(node) max by(namespace, pod, node) (kube_pod_info)
+      record: node_namespace_pod_container:container_memory_swap
```
Comment (karancode) [inline manifests/prometheus-rules.yaml]: generated after `make generate-in-docker`, not specific to my changes.

File: manifests/grafana-dashboardDefinitions.yaml [hunk 2]
```diff
@@ -11020,6 +11020,32 @@ items:
...
+
+                      ],
+                      "tagsQuery": "",
+                      "type": "query",
+                      "useTags": false
+                  },
+                  {
+                      "allValue": null,
+                      "current": {
+
```
Comment (karancode) [inline manifests/grafana-dashboardDefinitions.yaml]: all the changes in this file are generated after make `generate-in-docker`, not specific my to changes.

Comment (karancode) [reply manifests/grafana-dashboardDefinitions.yaml]: this seems relatable to #269

File: jsonnet/kube-prometheus/prometheus/prometheus.libsonnet [hunk 3]
```diff
@@ -440,5 +440,37 @@ local k = import 'ksonnet/ksonnet.beta.4/k.libsonnet';
           ],
         },
       },
+    serviceMonitorAwsEksCNI:
```
Comment (brancz) [inline jsonnet/kube-prometheus/prometheus/prometheus.libsonnet]: this should not be in the default, only in the eks patch, like the service

File: examples/kustomize.jsonnet [hunk 4]
```diff
@@ -1,5 +1,6 @@
 local kp =
-  (import 'kube-prometheus/kube-prometheus.libsonnet') + {
+  (import 'kube-prometheus/kube-prometheus.libsonnet') + 
+  (import 'kube-prometheus/kube-prometheus-aws-eks-cni.libsonnet') + {
```
Comment (brancz) [inline examples/kustomize.jsonnet]: this example should be generic and not specific to eks

File: docs/EKS-cni-support.md [hunk 5]
```diff
@@ -0,0 +1,22 @@
+# CNI monitoring special configuration updates for EKS
+
+AWS EKS uses [CNI](https://github.com/aws/amazon-vpc-cni-k8s) networking plugin for pod networking in Kubernetes using Elastic Network Interfaces on AWS
+
+One fatal issue that can occur is that you run out of IP addresses in your eks cluster. (Generally happens due to error configs where pods keep scheduling).
+
+You can monitor the `awscni` using kube-promethus with : 
+```
```
Comment (brancz) [inline docs/EKS-cni-support.md]: We manage all examples in the examples directory and then embed them in docs via `embedmd`. I would suggest looking at some other docs to see an example of that. As a bonus all snippets in the examples directory get run on CI so we know that they in fact work :)

File: jsonnet/kube-prometheus/prometheus/prometheus.libsonnet (hunk 3 above)
Comment (karancode) [reply jsonnet/kube-prometheus/prometheus/prometheus.libsonnet]: cbbfa0cad5ca7e231f8a255e95255d75fb75612c moves the service monitor to patch.

File: examples/kustomize.jsonnet (hunk 4 above)
Comment (karancode) [reply examples/kustomize.jsonnet]: d32e859a11b154bfdd29c8341a0d39cdab499215 revert back to the generic example.

File: examples/kustomize.jsonnet [hunk 6]
```diff
@@ -1,5 +1,6 @@
 local kp =
-  (import 'kube-prometheus/kube-prometheus.libsonnet') + {
+  (import 'kube-prometheus/kube-prometheus.libsonnet') + {
```
Comment (brancz) [inline examples/kustomize.jsonnet]: let's revert this completely so there's no change here anymore, it looks like it's just whitespace

File: docs/EKS-cni-support.md [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+local kp = (import 'kube-prometheus/kube-prometheus.libsonnet') +
+    (import 'kube-prometheus/kube-prometheus-aws-eks-cni.libsonnet') +
```
Comment (brancz) [inline docs/EKS-cni-support.md]: I didn't quite realize this before, but as we don't have an eks patch to begin with, I think we can rename this to be the "catch all" eks patch, as I assume this is the same for every eks cluster and all eks users want this.

Comment (karancode) [reply docs/EKS-cni-support.md]: Hi @brancz sorry but I didn't quite get it.. do you mean to rename the patch file as `kube-prometheus-catch-all-eks.libsonnet` ?

Comment (brancz) [reply docs/EKS-cni-support.md]: just `kube-prometheus-eks.libsonnet` :)

File: examples/kustomize.jsonnet (hunk 6 above)
Comment (karancode) [reply examples/kustomize.jsonnet]: 79c670bcd052cee7d0dcaf786cf5ff0b5b797421 this reverts the `examples/kustomize.jsonnet` file completely.

File: docs/EKS-cni-support.md (hunk 7 above)
Comment (karancode) [reply docs/EKS-cni-support.md]: 13c114a72f4a7a24d2579dcd8aed15e76fd0b838 updated.

File: docs/EKS-cni-support.md (hunk 5 above)
Comment (karancode) [reply docs/EKS-cni-support.md]: tried with `embedmd` b03ff4f5931c82c4acaf49a43972d3cfdf94f01f edb327531d88d876b6b3532f2f759aefb1625280

Comment (karancode) [reply docs/EKS-cni-support.md]: fixed the bug with example..
3c4dbc52d9b7727439d1d8595791c351b1c391fc d4ba158f9baab0a89466f6405c2d937954582247 resolves the issues.

**Comment by karancode** [comment]
@brancz fixed all issues... please take a look!

File: docs/EKS-cni-support.md [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+  _config+:: {
+    namespace: 'monitoring',
+  },
+  prometheusRules+:: {
+    groups+: [
+      {
+        name: 'example-group',
+        rules: [
+          {
+            record: 'aws_eks_available_ip',
```
Comment (brancz) [inline docs/EKS-cni-support.md]: What speaks against having all eks clusters benefit from this? :)

As in why just put it in the example and not part of the patch?

Comment (karancode) [reply docs/EKS-cni-support.md]: @brancz sorry for being super late here.
a3ab6bd49b39158222eef7309e1a28fc17f1dc07 this adds the rule in the patch itself.
I am figuring out other metrics as well.. will add them in future (with more alerts, etc.)

Comment (brancz) [reply docs/EKS-cni-support.md]: Awesome! Once CI passes, I'm super happy to merge this! :)

**Comment by brancz** [comment]
Try running `make --always-make clean generate-in-docker` and commit the changes. I think that should resolve everything on CI.

File: docs/EKS-cni-support.md (hunk 8 above)
Comment (karancode) [reply docs/EKS-cni-support.md]: I am figuring out why it is failing.. it started to happen after I rebased my branch with master.. 🙄

**Comment by brancz** [comment]
Make sure you’re using the latest jsonnet-CI docker image. Sorry there were some upstream changes in jb. 

Run “docker pull quay.io/coreos/jsonnet-ci:latest” before rerunning the command in the previous comment. Sorry for the inconvenience!

**Comment by karancode** [comment]
@brancz thanks alot!
Worked with the latest image. PTAL. 🙇

**Comment by brancz** [comment]
Thank you so much and thanks for bearing with us! :slightly_smiling_face: :tada:
